upi/aws/cloudformation/02_cluster_infra.yaml

```diff
@@ -264,6 +264,80 @@ Resources:
       Runtime: "python3.7"
       Timeout: 120
 
+  RegisterSubnetTagsLambdaIamRole:
+    Type: AWS::IAM::Role
+    Properties:
+      RoleName: !Join ["-", [!Ref ClusterName, "subnet-tags-lambda-role"]]
+      AssumeRolePolicyDocument:
+        Version: "2012-10-17"
+        Statement:
+        - Effect: "Allow"
+          Principal:
+            Service:
+            - "lambda.amazonaws.com"
+          Action:
+          - "sts:AssumeRole"
+      Path: "/"
+      Policies:
+      - PolicyName: !Join ["-", [!Ref ClusterName, "subnet-tagging-policy"]]
+        PolicyDocument:
+          Version: "2012-10-17"
+          Statement:
+          - Effect: "Allow"
+            Action:
+              [
+                "ec2:DeleteTags",
+                "ec2:CreateTags"
+              ]
+            Resource: "arn:aws:ec2:*:*:subnet/*"
+          - Effect: "Allow"
+            Action:
+              [
+                "ec2:DescribeSubnets",
+                "ec2:DescribeTags"
+              ]
+            Resource: "*"
+
+  RegisterSubnetTags:
+    Type: "AWS::Lambda::Function"
+    Properties:
+      Handler: "index.handler"
+      Role:
+        Fn::GetAtt:
+        - "RegisterSubnetTagsLambdaIamRole"
+        - "Arn"
+      Code:
+        ZipFile: |
+          import json
+          import boto3
+          import cfnresponse
+          def handler(event, context):
+            ec2_client = boto3.client('ec2')
+            if event['RequestType'] == 'Delete':
+              for subnet_id in event['ResourceProperties']['Subnets']:
+                ec2_client.delete_tags(Resources=[subnet_id], Tags=[{'Key': 'kubernetes.io/cluster/' + event['ResourceProperties']['ClusterID']}]);
+            elif event['RequestType'] == 'Create':
+              for subnet_id in event['ResourceProperties']['Subnets']:
+                ec2_client.create_tags(Resources=[subnet_id], Tags=[{'Key': 'kubernetes.io/cluster/' + event['ResourceProperties']['ClusterID'], 'Value': 'shared'}]);
+            responseData = {}
+            cfnresponse.send(event, context, cfnresponse.SUCCESS, responseData, event['ResourceProperties']['ClusterID']+event['ResourceProperties']['Subnets'][0])
+      Runtime: "python3.7"
+      Timeout: 120
+
+  RegisterPublicSubnetTags:
```

Comment:
I am extremely new to CloudFormation, so apologies if this should be obvious, but how is `RegisterPublicSubnetTags` wired in?  And with the declarative `kubernetes.io/role/internal-elb` you have earlier in this commit, what is the purpose of this apparently-dynamic stuff?

Reply:
kubernetes.io/role/internal-elb is wired directly on the private subnets in 01_vpc because it is a universal tag (it's where the internal elbs go for any/all clusters).

RegisterPublicSubnetTags (and Private) is a custom resource. It calls to the lambda defined in the same template. The lambda is registered here in 02_cluster_infra to add the kubernetes.io = shared tags needed on the various subnets. It allows for sharing 1 VPC to many clusters (one of the requirement drivers for UPI). There isn't a built-in resource to allow associating tags to subnets after the original resource definition. These will add a =shared reference for each 02_cluster run against this VPC and also remove them cleanly when the stack is terminated.